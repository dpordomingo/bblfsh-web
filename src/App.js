import React, { Component } from 'react';
import styled from 'styled-components';
import 'normalize.css';
import SplitPane from 'react-split-pane';

import Header from './components/Header';
import Footer from './components/Footer';
import Spinner from './components/Spinner';
import Editor from './components/Editor';
import Options from './components/Options';
import UASTViewer from './components/UASTViewer';
import { Notifications, Error } from './components/Notifications';
import * as history from './services/history';
import * as api from './services/api';
import { connect } from 'react-redux';
import { remove as errorsRemove } from 'state/errors';
import { init } from 'state';

const Wrap = styled.div`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  display: flex;
  flex-direction: column;
`;

const Content = styled.div`
  display: flex;
  height: 100%;
  flex-direction: row;
  position: relative;
`;

const RightPanel = styled.div`
  display: flex;
  height: 100%;
  flex-direction: column;
  position: relative;
`;

function getGistState(content, gistUrl, selectedLanguage) {
  const gistParts = gistUrl.split('/');
  const filename = gistParts[gistParts.length - 1];
  const state = {
    ...getResetCodeState(content, filename),
    cleanGist: true,
    gistUrl,
  };

  if (selectedLanguage) {
    state.selectedLanguage = selectedLanguage;
    history.setLanguage(selectedLanguage);
  }

  history.setGist(gistUrl);

  return state;
}

function getResetCodeState(code, filename, selectedLanguage) {
  return {
    code: code || null,
    ast: {},
    dirty: true,
    cleanGist: false,
    loading: false,
    errors: [],

    filename,
    actualLanguage: '',
    selectedLanguage: selectedLanguage || 'auto',
    selectedExample: '',
  };
}

export class App extends Component {
  componentDidMount() {
    return this.props.init();
  }

  onGistLoaded(gistUrl, selectedLanguage) {
    api
      .getGist(gistUrl)
      .then(content => {
        this.setState(
          getGistState(content, gistUrl, selectedLanguage),
          this.onRunParser
        );
      })
      .catch(errors => this.setState({ errors }));
  }

  renderContent() {
    const { innerWidth: width } = window;

    return (
      <SplitPane
        split="vertical"
        minSize={width * 0.25}
        defaultSize="50%"
        maxSize={width * 0.75}
      >
        <Editor />

        <RightPanel>
          <Options />
          <UASTViewer />
        </RightPanel>
      </SplitPane>
    );
  }

  render() {
    const { code, errors, errorsRemove } = this.props;

    return (
      <Wrap>
        <Header />
        <Content>{code !== null ? this.renderContent() : <Spinner />}</Content>

        <Footer />

        {errors.length > 0 ? (
          <Notifications>
            {errors.map((err, i) => {
              return (
                <Error message={err} key={i} onRemove={() => errorsRemove(i)} />
              );
            })}
          </Notifications>
        ) : null}
      </Wrap>
    );
  }
}

const mapStateToProps = state => ({
  languages: state.languages,
  code: state.code.code,
  errors: state.errors,
});

const mapDispatchToProps = {
  init,
  errorsRemove,
};

export default connect(mapStateToProps, mapDispatchToProps)(App);
