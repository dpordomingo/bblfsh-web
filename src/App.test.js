import React from 'react';
import { App } from './App';
import 'jest-styled-components';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

async function renderApp(state) {
  fetch.mockResponse(
    JSON.stringify([
      { id: 'python', name: 'Python', url: 'python-driver' },
      { id: 'java', name: 'Java', url: 'java-driver' },
    ])
  );

  const wrapper = mount(<App />);
  if (state) {
    wrapper.setState(state);
  }

  const promise = wrapper.instance().loaded;
  if (promise) {
    await promise;
  }

  return wrapper.instance();
}

it('renders without crashing', () => {
  const wrapper = shallow(<App errors={[]} code="foo = 1" />);
  expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
});

it('render while loading', async () => {
  const wrapper = shallow(<App errors={[]} code={null} />);
  expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
});

it('render with errors', async () => {
  const wrapper = shallow(<App errors={['error1', 'error2']} code="foo = 1" />);
  expect(toJson(wrapper)).toMatchStyledComponentsSnapshot();
});
