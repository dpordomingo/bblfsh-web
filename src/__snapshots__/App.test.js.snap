// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render while loading 1`] = `
<styled.div>
  <Connect(Header) />
  <styled.div>
    <styled.div />
  </styled.div>
  <Footer />
</styled.div>
`;

exports[`render with errors 1`] = `
<styled.div>
  <Connect(Header) />
  <styled.div>
    <SplitPane
      allowResize={true}
      defaultSize="50%"
      maxSize={768}
      minSize={256}
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": 7,
            "osversion": "7",
            "safari": true,
            "webkit": true,
            "windows": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": true,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {
              "alignContent": true,
              "alignItems": true,
              "alignSelf": true,
              "animation": true,
              "animationDelay": true,
              "animationDirection": true,
              "animationDuration": true,
              "animationFillMode": true,
              "animationIterationCount": true,
              "animationName": true,
              "animationPlayState": true,
              "animationTimingFunction": true,
              "appearance": true,
              "backdropFilter": true,
              "backfaceVisibility": true,
              "boxDecorationBreak": true,
              "breakAfter": true,
              "breakBefore": true,
              "breakInside": true,
              "clipPath": true,
              "columnCount": true,
              "columnFill": true,
              "columnGap": true,
              "columnRule": true,
              "columnRuleColor": true,
              "columnRuleStyle": true,
              "columnRuleWidth": true,
              "columnSpan": true,
              "columnWidth": true,
              "columns": true,
              "filter": true,
              "flex": true,
              "flexBasis": true,
              "flexDirection": true,
              "flexFlow": true,
              "flexGrow": true,
              "flexShrink": true,
              "flexWrap": true,
              "flowFrom": true,
              "flowInto": true,
              "fontKerning": true,
              "hyphens": true,
              "justifyContent": true,
              "mask": true,
              "maskBorder": true,
              "maskBorderMode": true,
              "maskBorderOutset": true,
              "maskBorderRepeat": true,
              "maskBorderSlice": true,
              "maskBorderSource": true,
              "maskBorderWidth": true,
              "maskClip": true,
              "maskComposite": true,
              "maskImage": true,
              "maskMode": true,
              "maskOrigin": true,
              "maskPosition": true,
              "maskRepeat": true,
              "maskSize": true,
              "maskType": true,
              "order": true,
              "perspective": true,
              "perspectiveOrigin": true,
              "regionFragment": true,
              "scrollSnapCoordinate": true,
              "scrollSnapDestination": true,
              "scrollSnapPointsX": true,
              "scrollSnapPointsY": true,
              "scrollSnapType": true,
              "shapeImageMargin": true,
              "shapeImageOutside": true,
              "shapeImageThreshold": true,
              "textDecorationColor": true,
              "textDecorationLine": true,
              "textDecorationSkip": true,
              "textDecorationStyle": true,
              "transform": true,
              "transformOrigin": true,
              "transformOriginX": true,
              "transformOriginY": true,
              "transformOriginZ": true,
              "transformStyle": true,
              "userSelect": true,
            },
          },
          "_requiresPrefix": Object {
            "alignContent": true,
            "alignItems": true,
            "alignSelf": true,
            "animation": true,
            "animationDelay": true,
            "animationDirection": true,
            "animationDuration": true,
            "animationFillMode": true,
            "animationIterationCount": true,
            "animationName": true,
            "animationPlayState": true,
            "animationTimingFunction": true,
            "appearance": true,
            "backdropFilter": true,
            "backfaceVisibility": true,
            "boxDecorationBreak": true,
            "breakAfter": true,
            "breakBefore": true,
            "breakInside": true,
            "clipPath": true,
            "columnCount": true,
            "columnFill": true,
            "columnGap": true,
            "columnRule": true,
            "columnRuleColor": true,
            "columnRuleStyle": true,
            "columnRuleWidth": true,
            "columnSpan": true,
            "columnWidth": true,
            "columns": true,
            "filter": true,
            "flex": true,
            "flexBasis": true,
            "flexDirection": true,
            "flexFlow": true,
            "flexGrow": true,
            "flexShrink": true,
            "flexWrap": true,
            "flowFrom": true,
            "flowInto": true,
            "fontKerning": true,
            "hyphens": true,
            "justifyContent": true,
            "mask": true,
            "maskBorder": true,
            "maskBorderMode": true,
            "maskBorderOutset": true,
            "maskBorderRepeat": true,
            "maskBorderSlice": true,
            "maskBorderSource": true,
            "maskBorderWidth": true,
            "maskClip": true,
            "maskComposite": true,
            "maskImage": true,
            "maskMode": true,
            "maskOrigin": true,
            "maskPosition": true,
            "maskRepeat": true,
            "maskSize": true,
            "maskType": true,
            "order": true,
            "perspective": true,
            "perspectiveOrigin": true,
            "regionFragment": true,
            "scrollSnapCoordinate": true,
            "scrollSnapDestination": true,
            "scrollSnapPointsX": true,
            "scrollSnapPointsY": true,
            "scrollSnapType": true,
            "shapeImageMargin": true,
            "shapeImageOutside": true,
            "shapeImageThreshold": true,
            "textDecorationColor": true,
            "textDecorationLine": true,
            "textDecorationSkip": true,
            "textDecorationStyle": true,
            "transform": true,
            "transformOrigin": true,
            "transformOriginX": true,
            "transformOriginY": true,
            "transformOriginZ": true,
            "transformStyle": true,
            "userSelect": true,
          },
          "_userAgent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.2 (KHTML, like Gecko) Safari/537.2",
          "prefixedKeyframes": "-webkit-keyframes",
        }
      }
      primary="first"
      split="vertical"
    >
      <Connect(Editor) />
      <styled.div>
        <Connect(Options) />
        <Connect(UASTViewer) />
      </styled.div>
    </SplitPane>
  </styled.div>
  <Footer />
  <Notifications>
    <Error
      message="error1"
      onRemove={[Function]}
    />
    <Error
      message="error2"
      onRemove={[Function]}
    />
  </Notifications>
</styled.div>
`;

exports[`renders without crashing 1`] = `
<styled.div>
  <Connect(Header) />
  <styled.div>
    <SplitPane
      allowResize={true}
      defaultSize="50%"
      maxSize={768}
      minSize={256}
      prefixer={
        Prefixer {
          "_browserInfo": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "name": "Safari",
            "osVersion": 7,
            "osversion": "7",
            "safari": true,
            "webkit": true,
            "windows": true,
            "x": true,
          },
          "_hasPropsRequiringPrefix": true,
          "_keepUnprefixed": false,
          "_metaData": Object {
            "browserName": "safari",
            "browserVersion": 7,
            "cssPrefix": "-webkit-",
            "jsPrefix": "Webkit",
            "keepUnprefixed": false,
            "requiresPrefix": Object {
              "alignContent": true,
              "alignItems": true,
              "alignSelf": true,
              "animation": true,
              "animationDelay": true,
              "animationDirection": true,
              "animationDuration": true,
              "animationFillMode": true,
              "animationIterationCount": true,
              "animationName": true,
              "animationPlayState": true,
              "animationTimingFunction": true,
              "appearance": true,
              "backdropFilter": true,
              "backfaceVisibility": true,
              "boxDecorationBreak": true,
              "breakAfter": true,
              "breakBefore": true,
              "breakInside": true,
              "clipPath": true,
              "columnCount": true,
              "columnFill": true,
              "columnGap": true,
              "columnRule": true,
              "columnRuleColor": true,
              "columnRuleStyle": true,
              "columnRuleWidth": true,
              "columnSpan": true,
              "columnWidth": true,
              "columns": true,
              "filter": true,
              "flex": true,
              "flexBasis": true,
              "flexDirection": true,
              "flexFlow": true,
              "flexGrow": true,
              "flexShrink": true,
              "flexWrap": true,
              "flowFrom": true,
              "flowInto": true,
              "fontKerning": true,
              "hyphens": true,
              "justifyContent": true,
              "mask": true,
              "maskBorder": true,
              "maskBorderMode": true,
              "maskBorderOutset": true,
              "maskBorderRepeat": true,
              "maskBorderSlice": true,
              "maskBorderSource": true,
              "maskBorderWidth": true,
              "maskClip": true,
              "maskComposite": true,
              "maskImage": true,
              "maskMode": true,
              "maskOrigin": true,
              "maskPosition": true,
              "maskRepeat": true,
              "maskSize": true,
              "maskType": true,
              "order": true,
              "perspective": true,
              "perspectiveOrigin": true,
              "regionFragment": true,
              "scrollSnapCoordinate": true,
              "scrollSnapDestination": true,
              "scrollSnapPointsX": true,
              "scrollSnapPointsY": true,
              "scrollSnapType": true,
              "shapeImageMargin": true,
              "shapeImageOutside": true,
              "shapeImageThreshold": true,
              "textDecorationColor": true,
              "textDecorationLine": true,
              "textDecorationSkip": true,
              "textDecorationStyle": true,
              "transform": true,
              "transformOrigin": true,
              "transformOriginX": true,
              "transformOriginY": true,
              "transformOriginZ": true,
              "transformStyle": true,
              "userSelect": true,
            },
          },
          "_requiresPrefix": Object {
            "alignContent": true,
            "alignItems": true,
            "alignSelf": true,
            "animation": true,
            "animationDelay": true,
            "animationDirection": true,
            "animationDuration": true,
            "animationFillMode": true,
            "animationIterationCount": true,
            "animationName": true,
            "animationPlayState": true,
            "animationTimingFunction": true,
            "appearance": true,
            "backdropFilter": true,
            "backfaceVisibility": true,
            "boxDecorationBreak": true,
            "breakAfter": true,
            "breakBefore": true,
            "breakInside": true,
            "clipPath": true,
            "columnCount": true,
            "columnFill": true,
            "columnGap": true,
            "columnRule": true,
            "columnRuleColor": true,
            "columnRuleStyle": true,
            "columnRuleWidth": true,
            "columnSpan": true,
            "columnWidth": true,
            "columns": true,
            "filter": true,
            "flex": true,
            "flexBasis": true,
            "flexDirection": true,
            "flexFlow": true,
            "flexGrow": true,
            "flexShrink": true,
            "flexWrap": true,
            "flowFrom": true,
            "flowInto": true,
            "fontKerning": true,
            "hyphens": true,
            "justifyContent": true,
            "mask": true,
            "maskBorder": true,
            "maskBorderMode": true,
            "maskBorderOutset": true,
            "maskBorderRepeat": true,
            "maskBorderSlice": true,
            "maskBorderSource": true,
            "maskBorderWidth": true,
            "maskClip": true,
            "maskComposite": true,
            "maskImage": true,
            "maskMode": true,
            "maskOrigin": true,
            "maskPosition": true,
            "maskRepeat": true,
            "maskSize": true,
            "maskType": true,
            "order": true,
            "perspective": true,
            "perspectiveOrigin": true,
            "regionFragment": true,
            "scrollSnapCoordinate": true,
            "scrollSnapDestination": true,
            "scrollSnapPointsX": true,
            "scrollSnapPointsY": true,
            "scrollSnapType": true,
            "shapeImageMargin": true,
            "shapeImageOutside": true,
            "shapeImageThreshold": true,
            "textDecorationColor": true,
            "textDecorationLine": true,
            "textDecorationSkip": true,
            "textDecorationStyle": true,
            "transform": true,
            "transformOrigin": true,
            "transformOriginX": true,
            "transformOriginY": true,
            "transformOriginZ": true,
            "transformStyle": true,
            "userSelect": true,
          },
          "_userAgent": "Mozilla/5.0 (Windows NT 6.1) AppleWebKit/537.2 (KHTML, like Gecko) Safari/537.2",
          "prefixedKeyframes": "-webkit-keyframes",
        }
      }
      primary="first"
      split="vertical"
    >
      <Connect(Editor) />
      <styled.div>
        <Connect(Options) />
        <Connect(UASTViewer) />
      </styled.div>
    </SplitPane>
  </styled.div>
  <Footer />
</styled.div>
`;
